File: C#/Socketry/socketry/SocketryServer.cs
using packetparser;
using socket;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace socketry
{
    public class SocketryServer : Socketry
    {
        private static short LINK_PORT_STATUS = (short)10000;

        public SocketryServer(int serverPort, Dictionary<String, Func<byte[], byte[]>> _procedures)
        {
            this.SetProcedures(_procedures);

            IServerSocket initServer = StartAt(serverPort);
            if (initServer == null) { 
                // throw exception
            }
            initServer.ConfigureBlocking(true);

            ISocket clientInitChannel = initServer.Accept();

            Link link = new Link(clientInitChannel);
            link.ConfigureBlocking(true);

            Packet initPacket = link.GetPacket();

            if(!(initPacket is not Packet.Init))
            {
                Console.WriteLine("Expected init packet...");
                //throw exception
            }

            byte[] socketsPerTunnel = ((Packet.Init)initPacket).channels;

            List<IServerSocket> serverSockets = new List<IServerSocket> ();
            List<short> ports = new List<short> ();
            short currentPort = LINK_PORT_STATUS;

            foreach(byte socketNum in socketsPerTunnel)
            {
                for(short i = 0; i < socketNum; i++)
                {
                    while (true)
                    {
                        currentPort++;
                        IServerSocket serverSocket = StartAt(currentPort);
                        if (serverSocket == null) { 
                            continue;
                        }

                        serverSockets.Add(serverSocket);
                        ports.Add(currentPort);
                        break;
                    }
                }
            }

            short[] portsArray = new short[ports.Count];
            for(int i = 0; i < ports.Count; i++)
            {
                portsArray[i] = ports[i];
            }

            Packet acceptPacket = new Packet.Accept(portsArray);
            link.SendPacket(acceptPacket);

            List<ISocket> clientSockets = new List<ISocket> ();

            foreach (IServerSocket serverSocket in serverSockets)
            {
                serverSocket.ConfigureBlocking (true);
                ISocket clientChannel = serverSocket.Accept();
                clientSockets.Add(clientChannel);
            }

            this.SetTunnelsFromSockets(clientSockets.ToArray(), socketsPerTunnel);

        }

        public void SetTunnelsFromSockets(ISocket[] sockets, byte[] socketsPerTunnel)
        {
            List<Tunnel> tunnels = new List<Tunnel>();

            byte LastSocketNum = 0;
            foreach (byte socketNum in socketsPerTunnel)
            {
                List<ISocket> socketsForTunnel = sockets.Skip(LastSocketNum).Take(socketNum).ToList();
                LastSocketNum += socketNum;
                try
                {
                    Tunnel tunnel = new Tunnel(socketsForTunnel.ToArray<ISocket>());
                    tunnels.Add(tunnel);
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.ToString());
                }
            }
        }

        private IServerSocket StartAt(int serverPort)
        {
            IServerSocket serverSocketChannel;
            try
            {
                serverSocketChannel = new ServerSocketTCP();
                serverSocketChannel.Bind(new IPEndPoint(IPAddress.Loopback, serverPort));
                return serverSocketChannel;
            }
            catch (Exception e) { 
                Console.WriteLine(e.ToString());
            }
            return null;
        }
    }
}

File: JAVA/socketry-main/src/main/java/com/socketry/SocketryServer.java
package com.socketry;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.concurrent.ExecutionException;
import java.util.function.Function;

import com.socketry.packetparser.Packet;
import com.socketry.socket.IServerSocket;
import com.socketry.socket.ISocket;
import com.socketry.socket.ServerSocketTCP;

public class SocketryServer extends Socketry {
    private static final short LINK_PORT_START = (short) 10000;

    /**
     * Opens a server and wait for client to connect
     * @param serverPort
     * @throws IOException
     * @throws InterruptedException
     * @throws ExecutionException
     */
    public SocketryServer(int serverPort, HashMap<String, Function<byte[], byte[]>> _procedures)
        throws IOException, InterruptedException, ExecutionException {

        this.setProcedures(_procedures);
        
        IServerSocket initServer = startAt(serverPort);
        if (initServer == null) {
            throw new IOException("Unable to start server");
        }
        initServer.configureBlocking(true);

        ISocket clientInitChannel = initServer.accept();
        
        // block to complete the handshake
        Link link = new Link(clientInitChannel);
        link.configureBlocking(true);

        Packet initPacket = link.getPacket();

        if (!(initPacket instanceof Packet.Init)) {
            throw new IllegalStateException("Expected Init packet" + initPacket);
        }

        byte[] socketsPerTunnel = ((Packet.Init) initPacket).channels();
        ArrayList<IServerSocket> serverSockets = new ArrayList<>();
        ArrayList<Short> ports = new ArrayList<>();
        short current_port = LINK_PORT_START;

        for (byte socketNum : socketsPerTunnel) {
            for (short i = 0; i < socketNum; i++) {
                while (true) {
                    current_port++;
                    IServerSocket socketChannel = startAt(current_port);
                    if (socketChannel == null) {
                        continue;
                    }
                    serverSockets.add(socketChannel);
                    ports.add(current_port);
                    break;
                }
            }
        }

        short[] portsArray = new short[ports.size()];
        for (int i = 0; i < ports.size(); i++) {
            portsArray[i] = ports.get(i);
        }

        Packet acceptPacket = new Packet.Accept(portsArray);
        link.sendPacket(acceptPacket);

        ArrayList<ISocket> clientSockets = new ArrayList<>();

        for (IServerSocket serverSocket : serverSockets) {
            serverSocket.configureBlocking(true);
            ISocket clientChannel = serverSocket.accept();
            clientSockets.add(clientChannel);
        }

        this.setTunnelsFromSockets(clientSockets.toArray(new ISocket[0]), socketsPerTunnel);
    }

    public void setTunnelsFromSockets(ISocket[] sockets, byte[] socketsPerTunnel) {
        ArrayList<Tunnel> tunnels = new ArrayList<>();

        byte lastSocketNum = 0;
        for (byte socketsNum: socketsPerTunnel) {
            ArrayList<ISocket> socketsForTunnel =
                new ArrayList<>(Arrays.asList(sockets).subList(lastSocketNum, lastSocketNum + socketsNum));
            lastSocketNum += socketsNum;
            try {
                Tunnel tunnel = new Tunnel(socketsForTunnel.toArray(new ISocket[0]));
                tunnels.add(tunnel);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        this.tunnels = tunnels.toArray(new Tunnel[0]);
    }

    private IServerSocket startAt(int serverPort) {
        IServerSocket serverSocketChannel;
        try {
            serverSocketChannel = new ServerSocketTCP();
            serverSocketChannel.bind(new InetSocketAddress(serverPort));
            return serverSocketChannel;
        } catch (IOException e) {
            System.out.println("Message: " + e.getMessage());
            e.printStackTrace();
        }
        return null;
    }
}
